### DISCUSSION TITLE
How to avoid queued job with cron schedule?

### DISCUSSION THREAD

**Discussion by seanmars** [quartznet/quartznet#1805] [Q&A]
**Describe the bug**

I have a job that takes 10 seconds, but I use corn schedule every 5 seconds to trigger that job.
But when the job is done, the next job will fire immediately.

**Version used**

Quartz.Net version: 3.5.0

```
.NET SDK (reflecting any global.json):
 Version:   6.0.402
 Commit:    6862418796

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19044
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\6.0.402\

global.json file:
  Not found

Host:
  Version:      6.0.10
  Architecture: x64
  Commit:       5a400c212a
```

**To Reproduce**

```cs
builder.Services.AddQuartz(q =>
{
    q.SchedulerId = "Scheduler-Core";

    q.UseMicrosoftDependencyInjectionJobFactory();

    // these are the defaults
    q.UseSimpleTypeLoader();
    q.UseInMemoryStore();
    q.UseDefaultThreadPool(tp =>
    {
        tp.MaxConcurrency = 1;
    });

    q.AddJob<AnalysisJob>(j => j
        .WithIdentity("analysis", "job1")
        .StoreDurably(false)
        .PersistJobDataAfterExecution(false)
        .WithDescription("analysis job")
    );

    q.AddTrigger(t => t.WithIdentity("Analysis Trigger")
        .ForJob("analysis", "job1")
        .StartNow()
        .WithCronSchedule("*/5 * * * * ?", scheduleBuilder =>
        {
            scheduleBuilder.WithMisfireHandlingInstructionDoNothing();
        })
    );
});

// ASP.NET Core hosting
builder.Services.AddQuartzServer(options =>
{
    // when shutting down we want jobs to complete gracefully
    options.WaitForJobsToComplete = true;
});
```

AnalysisJob.cs

```cs
using Quartz;

namespace JobSchedule.Jobs;

[DisallowConcurrentExecution]
public class AnalysisJob : IJob
{
    private readonly ILogger<AnalysisJob> _logger;

    public AnalysisJob(ILogger<AnalysisJob> logger)
    {
        _logger = logger;
    }

    public async Task Execute(IJobExecutionContext context)
    {
        _logger.LogInformation("{Name} running...", nameof(AnalysisJob));

        await Task.Delay(7 * 1000);

        _logger.LogInformation("{Name} done", nameof(AnalysisJob));
    }
}
```

**Expected behavior**

I want to avoid the job being queued.
10:10:05 fire the job
10:10:12 job done
10:10:15 fire the next job

**Additional context**

**Comment by Keen13** [comment]
> scheduleBuilder.WithMisfireHandlingInstructionDoNothing();

That's exactly how DoNothing misfire instruction works 

            /// <summary>
            /// Instructs the <see cref="IScheduler" /> that upon a mis-fire
            /// situation, the <see cref="ICronTrigger" /> wants to have it's
            /// next-fire-time updated to the next time in the schedule after the
            /// current time (taking into account any associated <see cref="ICalendar" />),
            /// but it does not want to be fired now.
            /// </summary>
            public const int DoNothing = 2;
            
For your expected behaviour you should use WithMisfireHandlingInstructionFireAndProceed() instruction, which is default for cron triggers, btw.

**Reply by seanmars** [reply]
Use `WithMisfireHandlingInstructionFireAndProceed` is not what I need.
If use `WithMisfireHandlingInstructionFireAndProceed` the process will be:

```
[2022-11-08 11:12::55] info: JobSchedule.Jobs.AnalysisJob[0] AnalysisJob start >>> 1 start
[2022-11-08 11:13::00] info: JobSchedule.Jobs.AnalysisJob[0] AnalysisJob start >>> 2 start
[2022-11-08 11:13::02] info: JobSchedule.Jobs.AnalysisJob[0] AnalysisJob done >>> 1 end
[2022-11-08 11:13::05] info: JobSchedule.Jobs.AnalysisJob[0] AnalysisJob start >>> 3 start
[2022-11-08 11:13::07] info: JobSchedule.Jobs.AnalysisJob[0] AnalysisJob done >>> 2 end
...
...
...
```

But what I want is:

```
[2022-11-08 11:12::55] info: JobSchedule.Jobs.AnalysisJob[0] AnalysisJob start >>> 1 start
[2022-11-08 11:13::00] >>> don't fire
[2022-11-08 11:13::02] info: JobSchedule.Jobs.AnalysisJob[0] AnalysisJob done >>> 1 end
[2022-11-08 11:13::05] info: JobSchedule.Jobs.AnalysisJob[0] AnalysisJob start >>> 2 start
[2022-11-08 11:13::10] >>> don't fire
[2022-11-08 11:13::12] info: JobSchedule.Jobs.AnalysisJob[0] AnalysisJob done >>> 2 end
```

**Comment by drusellers** [comment]
Could you have the job reschedule itself for 5 seconds in the future at the end of the job?

```
[2022-11-08 11:12::55] info: JobSchedule.Jobs.AnalysisJob[0] AnalysisJob start >>> 1 start
[2022-11-08 11:13::02] info: JobSchedule.Jobs.AnalysisJob[0] AnalysisJob done >>> 1 end
// wait's 5 seconds
[2022-11-08 11:13::07] info: JobSchedule.Jobs.AnalysisJob[0] AnalysisJob start >>> 2 start
[2022-11-08 11:13::14] info: JobSchedule.Jobs.AnalysisJob[0] AnalysisJob done >>> 2 end
```

**Reply by seanmars** [reply]
Not good, but if there has no other way to do what I need, rescheduling is ok too.

**Reply by drusellers** [reply]
you could have the job see if its running already and if so, exit. 

If you can share more about the business problem of running two at a time I'm happy to try and help with some other approach.

**Comment by lahma** [comment]
Generally when problem statement is something like "I want to run something every 5 minutes" or once in an hour, I'd urge you to also check simple trigger. I know it's not as fancy as cron, but it's quite ideal for running things at fixed intervals without worry about the time zones. Especially if you check the [misfire instructions available](https://github.com/quartznet/quartznet/blob/main/src/Quartz/MisfireInstruction.cs), simple trigger has more than others due to its.... simplicity.

Cron is great for "run at 2PM, at given time zone", but this sounds like fixed interval triggering to me.
